### DISCUSSION TITLE
AutoFixture.AutoNSubstitute: System.InvalidCastException when mocking a class with type parameters

### DISCUSSION THREAD

**Discussion by crhistianramirez** [AutoFixture/AutoFixture#1242] [Q&A] [answered]
There is a class in a library I'm using that looks like this:

```csharp
public class Order<TContact> : Order where TContact : Contact
{
    // ...other properties
    public TContact Contact { get; set; }
}
```

This allows a user of that library to define an order with their own `Contact `  class that may have additional properties so long as it extends the base `Contact ` model. For example:

```csharp

public class MyOrder: Order<MyContact>
{}

public class MyContact Contact
{
    public string MyProperty { get; set; }
}

var order = new MyOrder();
var myProperty = order.Contacts.MyProperty; // sweet!
```

However if I try to use auto mocking to get an instance of `MyOrder` I get 
> System.InvalidCastException : Unable to cast object of type 'Library.Models.Contact' to type 'MyApp.Models.MyContact'.

Is there anything I can do so that automocking works as expected and retrieves a member of type MyContact?

**Comment by aivascu** [comment]
Hello @crhistianramirez ,

So far the information you've provided is a bit vague.
Could you provide a minimum reproducible example, with a list of packages you're using (including  versions)?

Also please use the [Q&A Discussions](https://github.com/AutoFixture/AutoFixture/discussions/categories/q-a) section for questions.

Moving question to discussions.

**Reply by crhistianramirez** [reply]
I've added a repro here: https://github.com/crhistianramirez/Autofixture1242Repro

Definitely seems to be something about the library works because the example I posted above actually works. If anybody can provide guidance it would be much appreciated

[Here is the model](https://github.com/ordercloud-api/ordercloud-dotnet-sdk/blob/master/src/OrderCloud.SDK/Generated/Models.cs#L910-L993) defined by the library

**Reply by crhistianramirez** [reply]
I've updated the repro to extract only the most relevant bits from the library and was able to remove the library dependency. If I set a breakpoint on the setter for xp I can see autofixture correctly sets order.xp but then after that it gets set to an object for some reason. I'm still not clear why that's happening on autofixture side, trying to grok the source code

**Reply by crhistianramirez** [reply]
I think I got to the bottom of it. It looks like "xp" is defined twice - once on the base class and once on the derived class. The derived class hides it (purposely) but autofixture still tries to mock the base hidden member which causes issues. 

Seems like a bug to me?

**Comment by aivascu** [comment, marked as answer]
@crhistianramirez I have looked through the code and indeed there's a bug but it's not on AutoFixture's side. Your code violates the LSP principle. 
Consider the following scenario, you cast a `MyOrder` instance to `Order` and set the `dynamic xp` property to anything let's say `5`, then cast the object back to `MyOrder` and use the getter on `T xp` property, the expectation would be that when I use the instance as `MyOrder` everything will be fine, but cannot be fine since it is backed by the same dictionary instance and your implementation of `Order` does not handle this scenario.
AutoFixture cannot ignore the hidden properties, since the hidden property values can always be used, while its ower instance is casted to it's base type.

**Reply by crhistianramirez** [reply]
@aivascu  Thank you for taking the time to review the code, I really appreciate it.  I have a better understanding of the issue and why it can not and should not be handled on autofixture side. Cheers
